src/coprocessor/dag/batch/executors/top_n_executor.rs

```diff
@@ -0,0 +1,725 @@
+                        match_template_evaluable! {
+                            TT, match (src_value, result[column_idx].mut_decoded()) {
+                                (VectorValue::TT(src), VectorValue::TT(dest)) => {
+                                    dest.push(src[item.row_index].clone());
+                                },
+                                _ => unreachable!(),
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        result
+    }
+}
+
+impl<Src: BatchExecutor> BatchExecutor for BatchTopNExecutor<Src> {
+    #[inline]
+    fn schema(&self) -> &[FieldType] {
+        self.src.schema()
+    }
+
+    #[inline]
+    fn next_batch(&mut self, _scan_rows: usize) -> BatchExecuteResult {
+        assert!(!self.is_ended);
+
+        if self.n == 0 {
+            self.is_ended = true;
+            return BatchExecuteResult {
+                data: LazyBatchColumnVec::empty(),
+                warnings: EvalWarnings::default(),
+                is_drained: Ok(true),
+            };
+        }
+
+        let result = self.handle_next_batch();
+
+        match result {
+            Err(e) => {
+                // When there are error, we can just return empty data.
+                self.is_ended = true;
+                BatchExecuteResult {
+                    data: LazyBatchColumnVec::empty(),
+                    warnings: self.context.take_warnings(),
+                    is_drained: Err(e),
+                }
+            }
+            Ok(Some(data)) => {
+                self.is_ended = true;
+                BatchExecuteResult {
+                    data,
+                    warnings: self.context.take_warnings(),
+                    is_drained: Ok(true),
+                }
+            }
+            Ok(None) => BatchExecuteResult {
+                data: LazyBatchColumnVec::empty(),
+                warnings: self.context.take_warnings(),
+                is_drained: Ok(false),
+            },
+        }
+    }
+
+    #[inline]
+    fn collect_statistics(&mut self, destination: &mut BatchExecuteStatistics) {
+        self.src.collect_statistics(destination);
+    }
+}
+
+/// The item in the heap of `BatchTopNExecutor`.
+///
+/// WARN: The content of this structure is valid only if `BatchTopNExecutor` is valid (i.e.
+/// not dropped). Thus it is called unsafe.
+pub struct HeapItemUnsafe {
+    /// A pointer to the `order_is_desc` field in `BatchTopNExecutor`.
+    order_is_desc_ptr: NonNull<Pin<Vec<bool>>>,
+    /// The source columns that evaluated column in this structure is referring to.
+    ///
+    /// Multiple `HeapItemUnsafe` may share the same source column. Thus source columns
+    /// are placed behind a reference counter. However, there won't be a place other than
+    /// `HeapItemUnsafe` holding this reference counter, so Rc won't break
+    /// `BatchTopNExecutor: Send`.
+    source_column_data: Pin<Rc<LazyBatchColumnVec>>,
```

Comment:
For reviewers: This character needs to be verified carefully. Please help confirm this Rc is correct. @sticnarf @lonng

Reply:
There is a risk. `BinaryHeap` is from the standard library. We cannot make sure that `BinaryHeap` won't move the `HeapItemUnsafe` to other threads. (Although it's unreasonable for `BinaryHeap` to do so.

Reply:
If then, we might get data races when multiple `Rc`s get dropped at the same time.

Reply:
Sorry, not actually a problem

Reply:
Yes, not actually a problem since `HeapItemUnsafe` remains `!Sync + !Send`.